File: mathewDavidov/assignments/fastSlow/lc876/lc876.java
/**
 * Problem Statement:
 * Given a non-empty, singly linked list with head node head, return a middle node of a linked list.
 * If there are two middle nodes, return the second middle node.
 * 
 * Example:
 * Input: [1,2,3,4,5]
 * Output: Node 3 from this list (Serialization: [3,4,5])
 * The returned node has value 3.  (The judge's serialization of this node is [3,4,5]).
 * Note that we returned a ListNode object ans, such that:
 * ans.val = 3, ans.next.val = 4, ans.next.next.val = 5, and ans.next.next.next = NULL.
 * 
 * Input: [1,2,3,4,5,6]
 * Output: Node 4 from this list (Serialization: [4,5,6])
 * Since the list has two middle nodes with values 3 and 4, we return the second one.
 * 
 * [1,2,3,4,5,6]
 * fast - 1 -> 3 -> 5 -> null
 * slow - 1 -> 2 -> 3 -> 4
 * 
 * Edge cases:
 * Null node -> null
 * One node -> return the node (it's the middle)
 * 
 * initialize a slow pointer that references the head
 * initialize a fast pointer that references the head
 * loop while fast isn't null or its next isn't null:
 *   change fast to its nexts next pointer
 *   change slow to its next pointer
 * return the slow pointer
 * 
 * Runtime:
 * Time: O(n)
 * Space: O(1)
 */

class ListNode {
    public int value = 0;
    ListNode next;

    ListNode(int value) {
        this.value = value;
        next = null;
    }
};

public class lc876 {
    public static ListNode middleNode(ListNode head) {
        if (head == null || head.next == null) {
            return head;
        }

        ListNode fast = head, slow = head;

        while (fast != null && fast.next != null) {
            fast = fast.next.next;
            slow = slow.next;
        }

        return slow;
    }

    public static void main(String[] args) {
        ListNode head = new ListNode(1);
        head.next = new ListNode(2);
        head.next.next = new ListNode(3);
        head.next.next.next = new ListNode(4);
        head.next.next.next.next = new ListNode(5);
        System.out.println("Middle Node: " + middleNode(head).value);
      
        head.next.next.next.next.next = new ListNode(6);
        System.out.println("Middle Node: " + middleNode(head).value);
      
        head.next.next.next.next.next.next = new ListNode(7);
        System.out.println("Middle Node: " + middleNode(head).value);
    }
}

File: eliHenderson/assignments/fastSlow/lc876.js
/**
 * Summary: 
 *  If there is no cycle in the linked list 
 *  there will be a null at the of list
 *  we can use two pointer to iterate through the 
 *  array. One pointer will move 2x faster than the 
 *  other, Once the fast pointer reaches the end, 
 *  the slower pointer will be a the middle
 *  return the slow pointer
 * 
 * while (fast is not null)
 *  fast = fast.next.next
 *  slow = slow.next
 * 
 * return slow
 * 
 * Time: (O(N))
 * Space: O(1)

 */

class Node {
  constructor(value, next = null) {
    this.value = value;
    this.next = next;
  }
}

const find_middle_of_linked_list = function (head) {
  let hare = head;
  let turtle = head;
  while (hare !== null) {
    hare = hare.next;
    if (hare === null) break;
    hare = hare.next;

    turtle = turtle.next;
  }

  return turtle;
};

const find_middle_of_linked_list = function (head) {
  let hare = head;
  let turtle = head;

  while (hare !== null) {
    hare = hare.next;
    if (hare === null) break;
    hare = hare.next;

    turtle = turtle.next;
  }

  return turtle;
};

const find_middle_of_linked_list2 = (head) => {
  let hare = head;
  let turtle = head;

  while (hare !== null) {
    hare = hare.next.next;
    turtle = turtle.next;
  }

  return turtle;
};

const find_middle_of_linked_list = (head) => {
  let hare = head;
  let turtle = head;

  while (hare !== null && hare.next !== null) {
    hare = hare.next.next;
    turtle = turtle.next;
  }
  return turtle;
};

head = new Node(1);
head.next = new Node(2);
head.next.next = new Node(3);
head.next.next.next = new Node(4);
head.next.next.next.next = new Node(5);

console.log(`Middle Node: ${find_middle_of_linked_list(head).value}`);

head.next.next.next.next.next = new Node(6);
console.log(`Middle Node: ${find_middle_of_linked_list(head).value}`);

head.next.next.next.next.next.next = new Node(7);
console.log(`Middle Node: ${find_middle_of_linked_list(head).value}`);
